[Keyboard] Add FancyAlice66

chent7 opened the pull request:
<!--- Provide a general summary of your changes in the title above. -->
Added FancyAlice66 by Fancytech, with optional rotary encoder slot.
<!--- This template is entirely optional and can be removed, but is here to help both you and us. -->
<!--- Anything on lines wrapped in comments like these will not show up in the final text. -->

## Description

<!--- Describe your changes in detail here. -->

## Types of Changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply. -->
- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

drashna commented on keyboards/fancytech/fancyalice66/keymaps/via_encoder/keymap.c:
```diff
@@ -0,0 +1,79 @@
+bool encoder_update_user(uint8_t index, bool clockwise) {
+  if (index == 0) {
+    if (clockwise) {
+        tap_code(dynamic_keymap_get_keycode(biton32(layer_state), 4, 12));
```
This isn't the right way to do this. This limits it to basic keycodes, only.

Ideally, VIA should update to support the encoder map feature, as that natively support the dynamic keymap feature, etc... but that is highly unlikely.

chent7 replied:
I understand. Unfortunately Fancytech really wanted mapping support for VIA, that's the best I can do.

drashna commented on keyboards/fancytech/fancyalice66/keymaps/default_encoder/keymap.c:
```diff
@@ -0,0 +1,76 @@
+    )
+};
+
+#ifdef ENCODER_ENABLE
+bool encoder_update_user(uint8_t index, bool clockwise) {
+    if (clockwise) {
+      tap_code(KC_VOLU);
+    } else {
+      tap_code(KC_VOLD);
+    }
+    return false;
+}
+#endif
```
```suggestion
```

drashna commented on keyboards/fancytech/fancyalice66/fancyalice66.c:
```diff
@@ -0,0 +1,61 @@
+    }
+}
+
+#endif
```
This allows for a default behavior to be defined, without the need for the user to specify the behavior.  This is very useful for the configurator which doesn't have a way to configure this behavior, currently.

```suggestion
#endif

#ifdef ENCODER_ENABLE
bool encoder_update_kb(uint8_t index, bool clockwise) {
    if (!encoder_update_user(index, clockwise)) { 
        return false; 
    }
    if (index == 0) {
        if (clockwise) {
            tap_code_delay(KC_VOLU, 10);
        } else {
            tap_code_delay(KC_VOLD, 10);
        }
    }
    return true;
}
#endif
```

drashna reviewed (changes requested):
Also, if the only difference in the `encoder` keymaps are support for the encoders, then they should be removed.

chent7 commented:
I've removed the extra keymaps and added encoder support to kb.c as requested. I've managed to keep the encoder support for via, please let me know if that's acceptable

drashna commented on keyboards/fancytech/fancyalice66/keymaps/via/keymap.c:
```diff
@@ -0,0 +1,79 @@
+    )
+};
+
+#ifdef ENCODER_ENABLE
+bool encoder_update_user(uint8_t index, bool clockwise) {
+  if (index == 0) {
+    if (clockwise) {
+        tap_code(dynamic_keymap_get_keycode(biton32(layer_state), 4, 12));
+    } else {
+        tap_code(dynamic_keymap_get_keycode(biton32(layer_state), 4, 11));
+    }
+  }
+  return true;
+}
+#endif
```
This isn't going to work well, if at all. 

Eg, this only supports basic keycodes, so no rgb keycodes, no modded keycodes, etc. 

To be honest, you want something more like this: 
https://github.com/qmk/qmk_firmware/blob/8fa7b62011484f078548535556e91494a7f6cb8d/keyboards/work_louder/encoder_actions.c

This uses the "action_exec" path, which means that EVERY keycode, including custom keycodes are supported. (though certain keycodes, like mod tap and layer tap, and the like) won't work well).

However, ideally, via should be updated to support Encoder Maps, which are keymap like arrays, that support the underlying dynamic keymap functionality that VIA relies upon.

chent7 commented:
Please let me know if my implementation is acceptable

chent7 commented on keyboards/fancytech/fancyalice66/readme.md:
```diff
@@ -0,0 +1,27 @@
+# FancyAlice66
+
+![FancyAlice66](https://i.imgur.com/kg0TUlI.jpeg)
```
Thanks!

fauxpark commented on keyboards/fancytech/fancyalice66/info.json:
```diff
@@ -0,0 +1,83 @@
+{
```
This file is invalid and incorrect. Please use a JSON linter, and the Configurator to verify key position and ordering (press Ctrl+Shift+I and turn on Fast Input).

fauxpark replied:
You did not check it.
![image](https://user-images.githubusercontent.com/4781841/178453200-d78d45c1-6e73-413b-a599-359d9a4c8212.png)

chent7 replied:
Sorry I just assumed rotation would work and didn't verify it on the configurator. How do you put the json on the configurator? I also assume the layout needs to be flat now with no rotation correct?

fauxpark replied:
I've already told you how to preview the info.json in the Configurator. And yes, it does not support rotation right now.

chent7 replied:
Mind telling me where I can do that? Because it's not obvious.
![2022-07-12-1657617621](https://user-images.githubusercontent.com/34766434/178457109-7f7aad9b-8a17-49b0-b84a-af7edc52b864.png)

fauxpark replied:
Did you read my first comment?

chent7 replied:
Sorry I misread, anddd it's done now.

chent7 commented:
@fauxpark can you review the changes please

fauxpark commented on keyboards/fancytech/fancyalice66/config.h:
```diff
@@ -0,0 +1,87 @@
+
+#include "config_common.h"
+
+/* USB Device descriptor parameter */
+#define VENDOR_ID       0xECED
+#define PRODUCT_ID      0x3663
+#define DEVICE_VER      0x0001
+#define MANUFACTURER    Fancytech
+#define PRODUCT         FancyAlice66
+
```
```suggestion
```

fauxpark commented on keyboards/fancytech/fancyalice66/info.json:
```diff
@@ -0,0 +1,295 @@
+{
+	"keyboard_name": "FancyAlice66",
+	"url": "https://www.aliexpress.com/item/1005004272392611.html",
+	"maintainer": "chent7",
```
```suggestion
	"keyboard_name": "FancyAlice66",
	"manufacturer": "Fancytech",
	"url": "https://www.aliexpress.com/item/1005004272392611.html",
	"maintainer": "chent7",
	"usb": {
	    "vid": "0xECED",
	    "pid": "0x3663",
	    "device_version": "0.0.1"
	},
```

fauxpark commented on keyboards/fancytech/encoder_actions.c:
```diff
@@ -0,0 +1,68 @@
+#    ifdef ENCODERS
+    for (int index = 0; index < ENCODERS; ++index) {
+        if (encoder_state[index]) {
+            keyevent_t encoder_event = (keyevent_t) {
+                .key = encoder_state[index] >> 1 ? encoder_cw[index] : encoder_ccw[index],
+                .pressed = false,
+                .time = (timer_read() | 1)
+            };
+            encoder_state[index] = 0;
+            action_exec(encoder_event);
+        }
+    }
+#    endif
+}
+
+void encoder_action_register(uint8_t index, bool clockwise) {
+#    ifdef ENCODERS
+    keyevent_t encoder_event = (keyevent_t) {
+        .key = clockwise ? encoder_cw[index] : encoder_ccw[index],
+        .pressed = true,
+        .time = (timer_read() | 1)
+    };
+    encoder_state[index] = (clockwise ^ 1) | (clockwise << 1);
+    action_exec(encoder_event);
+#    endif
+}
+
+void matrix_scan_kb(void) {
+    encoder_action_unregister();
+    matrix_scan_user();
+}
+
+bool encoder_update_kb(uint8_t index, bool clockwise) {
+    encoder_action_register(index, clockwise);
+    // don't return user actions, because they are in the keymap
+    // encoder_update_user(index, clockwise);
+    return true;
+};
+
+#endif
```
This can be replaced with the new [Encoder Map](https://docs.qmk.fm/#/feature_encoders?id=encoder-map) feature.

ogabriel commented:
@chent7 any chance of openning another pull request? probably the maintainers didnt have time to review previosly, so it was automatically closed

chent7 commented:
> @chent7 any chance of openning another pull request? probably the maintainers didnt have time to review previosly, so it was automatically closed

I didn't have time to continue working on the new changes. But feel free to continue with new    PR if you would like to work on it.
